src/libraries/System.Private.CoreLib/src/System/UInt64.cs

```diff
@@ -176,6 +177,58 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
             return Number.TryParseBinaryInteger(s, style, NumberFormatInfo.GetInstance(provider), out result) == Number.ParsingStatus.OK;
         }
 
+        //
+        // Helpers, those methods are referenced from the JIT
+        //
+
+        [StackTraceHidden]
+        private static ulong MultiplyChecked(ulong left, ulong right)
```

Comment:
This seems like a very expensive way to implement `MultiplyChecked`.

For `UInt128` we simply do `BigMul` and check if `upper` is 0: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/UInt128.cs,1379

But more practically speaking, we can determine overflow just by doing `if ((128 - LeadingZeroCount(left) - LeadingZeroCount(right)) > 64)`. Since `x-bit * y-bit` will produce an `(x + y)-bit` product; anything else must fit into the given 64-bits.

Reply:
I'd prefer to change logic here in a separate PR. Keep in mind that this is for 32bit platforms so things like BigMul are slow there.

Reply:
> I'd prefer to change logic here in a separate PR

That's fine.

> Keep in mind that this is for 32bit platforms so things like BigMul are slow there.

That's why I suggested the extremely cheap check of `if ((128 - LeadingZeroCount(left) - LeadingZeroCount(right)) > 64)`. That, on a 32-bit platform, ends up being an almost free check to determine overflow without doing any expensive multiplication (it simply uses the `lzcnt` equivalent that every platform we target, including WASM, natively supports).

Then, you only have to do a regular 64x64-bit multiplication since you know it can't overflow. You don't have to do the extra work (and many branches) to check overflow at each step in the algorithm.

Reply:
> That, on a 32-bit platform, ends up being an almost free check to determine overflow without doing any expensive multiplication 

It looks like an approximate check to determine overflow. If it fails, you still need to take the more expensive multiplication to determine whether it is actually going to overflow - correct?

The current implementation has a variant of this idea where it checks whether the high 32-bits of both numbers are zero. If they are, it knows that there is not going to be an overflow and so it multiplies the low 32-bit portions and returns. If they are not, it does the more expensive multiplication.

I think the difference between what's there currently and your suggestion is in the speed of the fast path and the curoff for the fast path. The current implementation has lower cutoff and faster fast path. Your suggestion has higher cutoff and slower fast path.

Reply:
> It looks like an approximate check to determine overflow.

Yes.

> you still need to take the more expensive multiplication to determine whether it is actually going to overflow

Not quite.

--------------

In general, for an `x-bit * y-bit` value (excepting where one input is `0`), you need a minimum of `max(x, y)` bits to represent the product and a maximum of `x + y` bits.

The minimum bit count comes about because of the special case `1`, which returns the original value. Multiplication then adds one additional bit to the output for each power of 2 (multiplying by 2 is equivalent to shifting left by 1, multiplying by 4 is equivalent to shifting left by 2, thus multiplying by 3 must be between these two values and takes less than 2 additional bits).

Thus the exact number of bits needed is effectively `max(x, y) + ceil(log2(min(x, y)))` bits.

Finding the input with the smaller length and computing its effective `ceil(log2)` are both incredibly cheap operations that all our targets have direct acceleration to support. So it should always be significantly cheaper than what we're doing here already.

The approach I detailed of giving `if ((128 - LeadingZeroCount(left) - LeadingZeroCount(right)) > 64)` is an even cheaper approximation and there is some subset of valid inputs that would trigger it and for which we'd then need to compute the actual `max(x, y) + ceil(log2(min(x, y)))`.

So it basically just comes down to handling the 99% of inputs that can't possibly overflow as cheaply as possible. This is always fine because `lzcnt` is a native instruction for Arm, Arm64, and WASM. It is part of our default R2R target for x86/x64 and if you happen to be on x86/x64 that is pre-lzcnt (more than 11 years old at this point), you end up getting `(sizeof(T) * 8 - 1) ^ BSR` instead. For Arm32 and x86, you at worst get an extra branch and addition in here to account for the upper vs lower halves of the 64-bit input.

For the subset of cases that hit the "it could actually be under 64-bits", you then need to do `log2` which just uses the already computed `lzcnt` as input. So we remain with an incredibly efficient fast path that minimizes the total amount of work and doesn't need to do the extra multiplications required to check if bits `64-127` are non-zero.

Reply:
> So it basically just comes down to handling the 99% of inputs that can't possibly overflow as cheaply as possible. 

Right. It is 99% inputs from distribution found in real-world programs. The current implementation is optimized to be as cheap as possible for multiplying two <32-bit numbers. It is quite likely it aligns with the 99% inputs distribution found in real-world programs.

Reply:
This conversation made me check the perf of the rewrite. The rewrite clocks 1.25x slower on my quick lame micro-benchmark. Have you done https://github.com/dotnet/runtime/pull/98858#issuecomment-1986388593 for checked multiplication?

```
using System;
using System.Diagnostics;

for (;;)
{
    Stopwatch sw = new();
    sw.Start();
    Work(10_000_000_000);
    Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}

static ulong Work(ulong iters)
{
    ulong sum = 0;
    for (ulong i = 0; i < iters; i += 123)
    {
        sum += checked(i * 54321);
    }
    return sum;
}
```

Reply:
> Have you done [#98858 (comment)](https://github.com/dotnet/runtime/pull/98858#issuecomment-1986388593) for checked multiplication?

I haven't but here the code should be unchanged so this difference is probably from RyuJIT generating worse codegen than MSVC on X86.

Reply:
> Are you checking with Checked CoreLib btw? 

I always run perf tests on release builds of everything.

One of the problem seems to be that BigMul is not always inlined:

![image](https://github.com/dotnet/runtime/assets/6668460/d77a36ae-0a9e-4889-8ff1-8a092ff5ba4e)

It may be a good idea to split this PR into multiple smaller one so that each of the changes gets the right level of scrutiny.

Reply:
The existing code (both the C/C++ version in regular CoreCLR and the managed native AOT version) does not use the BigMul helper, so it does not have this perf bug.

Reply:
> The existing code (both the C/C++ version in regular CoreCLR and the managed native AOT version) does not use the BigMul helper, so it does not have this perf bug.

Putting `AggressiveInlining` on `BigMul` like I did in #100027 makes the regression be about 2.9% on CoreCLR on my machine, I'd argue that's good enough in exchange for sharing the implementation between runtimes. It will also likely improve with future optimizations to the JIT so I'd assume it's gonna get better than native in the future.

> It may be a good idea to split this PR into multiple smaller one so that each of the changes gets the right level of scrutiny.

While I'd agree with this before, I feel like splitting the PR at this point after most review was done would only complicate things here.

Reply:
> I feel like splitting the PR at this point after most review was done would only complicate things here.

It would allow me or other maintainers to get these changes merged with higher confidence and less work. In general, multiple smaller PRs are less total work for maintainers to get through than one big PR.

Reply:
I have peeled off first chunk into https://github.com/dotnet/runtime/pull/100222

Reply:
https://github.com/dotnet/runtime/pull/100222 is merged with number of adjustments. I think the next good-sized chunk to peel off into a separate PR is everything to do with the long multiplication. @MichalPetryka Are you interested in doing that?

Reply:
I'll look into splitting this up tomorrow.